bootstore/src/coordinator.rs

```diff
@@ -0,0 +1,427 @@
+                }
+            }
+        }
+
+        // Handle CommitOk messages
+        match result {
+            NodeOpResult::CommitOk { rack_uuid, epoch } => {
+                if rack_uuid != state.rack_uuid {
+                    return Err(Error::CommitOkBadRackUuid {
+                        from,
+                        expected: state.rack_uuid,
+                        actual: rack_uuid,
+                    });
+                }
+                if epoch != 0 {
+                    return Err(Error::CommitOkBadEpoch {
+                        from,
+                        expected: 0,
+                        actual: epoch,
+                    });
+                }
+                state.ackd_commits.insert(from);
+            }
+            NodeOpResult::PrepareOk { rack_uuid, epoch } => {
+                return Err(Error::UnexpectedPrepareOk {
+                    from,
+                    epoch,
+                    rack_uuid,
+                });
+            }
+            NodeOpResult::Share { epoch, .. } => {
+                return Err(Error::UnexpectedShare { from, epoch });
+            }
+        }
+
+        Ok(state.commit_complete())
+    }
+}
+
+/// A [`CoordinatorOperation`] for rack reconfiguration
+//
+// TODO: Not used yet
+#[allow(dead_code)]
+struct ReconfigureOperation {
+    old_epoch: i32,
+    old_members: Vec<Ed25519Certificate>,
+    new_epoch: i32,
+    new_members: Vec<Ed25519Certificate>,
+    share_distributions: BTreeMap<Ed25519Certificate, ShareDistribution>,
+}
+
+/// A coordinator for the bootstore's 2PC protocol
+///
+/// The coordinatore is responsible for creating the [`trust_quorum::RackSecret`]
+/// splitting it into share sand distributing those shares along with the
+/// relevant information to the participant nodes.
+pub struct Coordinator {
+    state: CoordinatorState,
+    op: Box<dyn CoordinatorOperation>,
+}
+
+impl Coordinator {
+    /// Create a coordinator used to initialize a rack
+    pub fn new_initialize(
+        log: Logger,
+        rack_uuid: Uuid,
+        members: BTreeSet<Ed25519Certificate>,
+    ) -> Result<Coordinator, Error> {
+        let log = log.new(o!(
+            "component" => "BootstoreCoordinator"
+        ));
+        let total_nodes = members.len();
+        let threshold = Self::threshold(total_nodes);
+        let secret = RackSecret::new();
+        let (shares, verifier) = secret
+            .split(threshold, total_nodes)
+            .map_err(Error::FailedToSplitSecret)?;
+        let share_distributions = members
+            .iter()
+            .cloned()
+            .zip(shares.into_iter().map(|share| ShareDistribution {
+                threshold,
+                verifier: verifier.clone(),
+                share,
+                member_device_id_certs: members.clone(),
+            }))
+            .collect();
+        let state = CoordinatorState {
+            id: 0,
```

Comment:
We don't in this case. By definition, the coordinator ID is always 0 for rack init since it's performed by RSS and there is no way to keep track of increments. If one RSS or scrimlet dies we just reuse `0`. This doesn't actually work if for some reason there is a laptop plugged into both scrimlets and initialization is attempted simultaneously. However, we prevent most issues there by preventing a commit of the wrong hashed prepare. So the commit stage won't succeed. There are still other isssues, but for this PR in particular I'm not worrying about them.

I'm also considering using a ULID for the coordinator ID, if we can ensure we have proper time via RSS. I think we can ensure that by using the time of the connected laptop. We know we'll have time during reconfiguration because the control plane will be up and Nexus will be doing the reconfiguring. The biggest problem with this solution is that the laptop could set some time so far in the future that nexus can't catch up and reconfiguration fails. More thought definitely needed here.